### ISSUE TITLE
pip install -e . error: subprocess-exited-with-error

### ISSUE THREAD

**Issue by xiongjiuli** [MIC-DKFZ/nnUNet#1866]
(nnunet2) [xxx@bme_login2 nnUNet]$ pip install -e .
Obtaining file:///xxx/data/xxx/nnUNet
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... error
  error: subprocess-exited-with-error

  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      error: Multiple top-level packages discovered in a flat-layout: ['log', 'nnunetv2', 'DataFrame'].

      To avoid accidental inclusion of unwanted files or directories,
      setuptools will not proceed with this build.

      If you are trying to create a single distribution with multiple packages
      on purpose, you should not rely on automatic discovery.
      Instead, consider the following options:

      1. set up custom discovery (`find` directive with `include` or `exclude`)
      2. use a `src-layout`
      3. explicitly set `py_modules` or `packages` with a list of names

      To find more information, look for "package discovery" on setuptools docs.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

**Comment by FabianIsensee** [comment]
Could it be that you have added additional stuff to the nnU-Net code? It seems like it is detecting multiple packages in the folder where the setup.py is located. Those are not present in the master

**Comment by FabianIsensee** [comment]
Hey, how are things going? Have you been able to install everything?

**Comment by xyjpang** [comment]
I encountered the same problem as well. Initially, after cloning the project, I executed pip install -e ., and there were no issues. However, after adding my own DATASET directory, when I tried to convert the dataset using the command nnUNet_convert_decathlon_task -i /root/siton-tmp/nnUNet/nnUNet/DATASET/nnUNet_raw/Task01_BrainTumour, I encountered an error: -bash: nnUNet_convert_decathlon_task: command not found. Following your previous advice, I re-executed pip install -e . only to encounter the error error: subprocess-exited-with-error. Even after removing the DATASET directory and re-executing pip install -e ., it still indicates that all packages have been successfully installed.

> Hey, how are things going? Have you been able to install everything?

**Comment by FabianIsensee** [comment]
Hey, since nnU-Net v2 the command is called `nnUNetv2_convert_MSD_dataset`. Can you please try that?

**Comment by FabianIsensee** [comment]
please DO NOT place downloaded MSD datasets into the nnUNet_raw folder

**Comment by xyjpang** [comment]
> Hey, since nnU-Net v2 the command is called `nnUNetv2_convert_MSD_dataset`. Can you please try that?
Yes, I realized that I used the wrong command. But I still encounter the following issue:

(base) root@bc8cc2d64fae:~# nnUNetv2_convert_MSD_dataset -i /root/siton-tmp/data/Task01_BrainTumour
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/opt/conda/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/opt/conda/lib/python3.10/multiprocessing/pool.py", line 51, in starmapstar
    return list(itertools.starmap(args[0], args[1]))
  File "/root/siton-tmp/nnUNet/nnUNet/nnunetv2/dataset_conversion/convert_MSD_dataset.py", line 15, in split_4d_nifti
    img_itk = sitk.ReadImage(filename)
  File "/opt/conda/lib/python3.10/site-packages/SimpleITK/extra.py", line 375, in ReadImage
    return reader.Execute()
  File "/opt/conda/lib/python3.10/site-packages/SimpleITK/SimpleITK.py", line 8430, in Execute
    return _SimpleITK.ImageFileReader_Execute(self)
RuntimeError: Exception thrown in SimpleITK ImageFileReader_Execute: /tmp/SimpleITK/Code/IO/src/sitkImageReaderBase.cxx:105:
sitk::ERROR: Unable to determine ImageIO reader for "/root/siton-tmp/data/Task01_BrainTumour/imagesTr/BRATS_001.nii.gz"
"""

**Comment by FabianIsensee** [comment]
This is strange, it appears like SImpleITK cannot read the images. Please verify that with

```
import SimpleITK as sitk
img = sitk.ReadImage("/root/siton-tmp/data/Task01_BrainTumour/imagesTr/BRATS_001.nii.gz")
```

Best,
Fabian

**Comment by FabianIsensee** [comment]
Have you been able to solve this?

**Comment by iTomxy** [comment]
**TL;DR** append the following to the `pyproject.toml` solves this problem (thanks ChatGPT):
```toml
[tool.setuptools]
packages = ["nnunetv2"]
```

---

I meet the same problem as @xiongjiuli. The thing is: I install nnUnet successfully once, and produced some other folders, with the current folder structure like:
```
nnUNet/
|- documentation/
|- nnunetv2/
|- LICENSE
|- pyproject.toml
|- readme.md
|- setup.py
|- __pycache__/    # newly added after the 1st successful installation from this down
|- data/
|- data_preprocessed/
|- results/
|- env.sh          # the same installing script
`- run.sh
```
The `env.sh` I used to install nnUNet is:
```shell
#!/bin/bash
set -e

CONDA_P=${1-"$HOME/miniconda3"}
CONDA_BIN=$CONDA_P/bin
ENV=cu118_pt251
if [ ! -d $CONDA_P/envs/$ENV ]; then
    $CONDA_BIN/conda create --name $ENV python=3.9 -y
fi
ENV_BIN=$CONDA_P/envs/$ENV/bin

$ENV_BIN/pip install torch==2.5.1 torchvision==0.20.1 --index-url https://download.pytorch.org/whl/cu118
$ENV_BIN/pip install nibabel itk simpleitk medpy # tensorboard opencv-python-headless
# conda install -n $ENV -y scikit-learn scipy matplotlib pandas Cython pyyaml scikit-image jupyter
$ENV_BIN/pip install -e .
$ENV_BIN/pip install blosc2
```
But I uninstalled it later because I removed the whole virtual environment containing it. The second time I install nnUNet with the same old `env.sh`, the full error is:
```
Obtaining file:///home/tom/codes/tmp.ptcloud/ribsegv2/segmentation/nnUNet
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Checking if build backend supports build_editable: started
  Checking if build backend supports build_editable: finished with status 'done'
  Getting requirements to build editable: started
  Getting requirements to build editable: finished with status 'error'
  error: subprocess-exited-with-error
  
  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> [54 lines of output]
      /tmp/pip-build-env-hgz5dr_g/overlay/lib/python3.9/site-packages/setuptools/config/_apply_pyprojecttoml.py:82: SetuptoolsDeprecationWarning: `project.license` as a TOML table is deprecated
      !!
      
              ********************************************************************************
              Please use a simple string containing a SPDX expression for `project.license`. You can also use `project.license-files`. (Both options available on setuptools>=77.0.0).
      
              By 2026-Feb-18, you need to update your project and remove deprecated calls
              or your builds will no longer be supported.
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************
      
      !!
        corresp(dist, value, root_dir)
      /tmp/pip-build-env-hgz5dr_g/overlay/lib/python3.9/site-packages/setuptools/config/_apply_pyprojecttoml.py:61: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!
      
              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:
      
              License :: OSI Approved :: Apache Software License
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************
      
      !!
        dist._finalize_license_expression()
      /tmp/pip-build-env-hgz5dr_g/overlay/lib/python3.9/site-packages/setuptools/dist.py:761: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!
      
              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:
      
              License :: OSI Approved :: Apache Software License
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************
      
      !!
        self._finalize_license_expression()
      error: Multiple top-level packages discovered in a flat-layout: ['data', 'results', 'nnunetv2', 'data_preprocessed'].
      
      To avoid accidental inclusion of unwanted files or directories,
      setuptools will not proceed with this build.
      
      If you are trying to create a single distribution with multiple packages
      on purpose, you should not rely on automatic discovery.
      Instead, consider the following options:
      
      1. set up custom discovery (`find` directive with `include` or `exclude`)
      2. use a `src-layout`
      3. explicitly set `py_modules` or `packages` with a list of names
      
      To find more information, look for "package discovery" on setuptools docs.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```
The main error is the same as @xiongjiuli:
```
error: Multiple top-level packages discovered in a flat-layout: ['data', 'results', 'nnunetv2', 'data_preprocessed'].
```
As pointed out by ChatGPT, the origin of this is those newly added folders other than `nnunetv2`, i.e. in my case: data/, results/, data_preprocessed/. So we have to tell the setup tool which folder is to be installed. So comes the solution in the top of this post.

I suggest @FabianIsensee add this to `pyproject.toml` for robustness.
